File: linux/richpresence/updater.py
"""Updates from GitHub releases. The AppImage replaces itself (after checking the file against the
SHA-256 GitHub publishes for it); the .deb and source installs only say a new version is out."""
import hashlib
import json
import os
import urllib.request
from pathlib import Path

from . import __version__

API = 'https://api.github.com/repos/noice912/RichPresence/releases/latest'
UA = {'User-Agent': 'RichPresence updater'}


def version_tuple(v):
    out = []
    for part in str(v).lstrip('v').split('.'):
        digits = ''.join(ch for ch in part if ch.isdigit())
        out.append(int(digits or 0))
    return tuple(out + [0] * (3 - len(out)))


def latest_release(url=API):
    req = urllib.request.Request(url, headers=UA)
    with urllib.request.urlopen(req, timeout=20) as r:
        return json.loads(r.read().decode('utf-8'))


def check(current=__version__, appimage=None, release=None, log=print):
    """Returns (state, latest_version, path): state is 'current', 'available' (can't self-install),
    'ready' (a verified new AppImage is at path) or 'failed'."""
    appimage = appimage if appimage is not None else os.environ.get('APPIMAGE')
    try:
        rel = release or latest_release()
        latest = str(rel.get('tag_name', '')).lstrip('v')
        if version_tuple(latest) <= version_tuple(current):
            return 'current', latest, None
        if not appimage:
            return 'available', latest, None
        asset = next((a for a in rel.get('assets', []) if a.get('name') == 'RichPresence-x86_64.AppImage'), None)
        want = str((asset or {}).get('digest') or '').replace('sha256:', '').lower()
        if not asset or not want:
            log(f'Update {latest} has no AppImage with a published checksum, so it was not installed.')
            return 'failed', latest, None
        target = Path(appimage)
        tmp = target.with_name(target.name + '.download')
        log(f'Downloading update {latest}...')
        h = hashlib.sha256()
        req = urllib.request.Request(asset['browser_download_url'], headers=UA)
        with urllib.request.urlopen(req, timeout=120) as r, open(tmp, 'wb') as f:
            while True:
                chunk = r.read(1 << 20)
                if not chunk:
                    break
                h.update(chunk)
                f.write(chunk)
        if h.hexdigest() != want:
            tmp.unlink(missing_ok=True)
            log(f"Update {latest} didn't match its checksum and was deleted.")
            return 'failed', latest, None
        tmp.chmod(0o755)
        log(f'Update {latest} downloaded and verified.')
        return 'ready', latest, str(tmp)
    except Exception as e:
        log(f"Couldn't check for updates: {e}")
        return 'failed', None, None


def install(downloaded, appimage=None):
    """Puts the new AppImage in place of the running one (Linux keeps the old file open for the
    running copy) and returns the path to start."""
    appimage = appimage or os.environ.get('APPIMAGE')
    os.replace(downloaded, appimage)
    return appimage

File: android/app/src/main/java/io/github/noice912/richpresence/Updater.kt
package io.github.noice912.richpresence

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.content.FileProvider
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest

/**
 * Updates from GitHub releases. Android always asks the person to confirm an install, so this
 * downloads the new APK, checks it against the SHA-256 GitHub publishes, and opens the installer.
 */
object Updater {
    private const val API = "https://api.github.com/repos/noice912/RichPresence/releases/latest"
    private const val ASSET = "RichPresence-android.apk"

    sealed class Result {
        object Current : Result()
        data class Ready(val version: String, val file: File) : Result()
        data class Failed(val why: String) : Result()
    }

    fun versionParts(v: String): List<Int> =
        v.trim().removePrefix("v").split('.').map { p -> p.takeWhile { it.isDigit() }.toIntOrNull() ?: 0 }
            .let { it + List((3 - it.size).coerceAtLeast(0)) { 0 } }

    fun isNewer(latest: String, current: String): Boolean {
        val a = versionParts(latest)
        val b = versionParts(current)
        for (i in 0 until maxOf(a.size, b.size)) {
            val x = a.getOrElse(i) { 0 }
            val y = b.getOrElse(i) { 0 }
            if (x != y) return x > y
        }
        return false
    }

    /** Runs on a background thread. */
    fun check(dir: File, current: String): Result = try {
        val rel = JSONObject(get(API).readBytes().toString(Charsets.UTF_8))
        val latest = rel.optString("tag_name").removePrefix("v")
        if (!isNewer(latest, current)) Result.Current
        else {
            val assets = rel.optJSONArray("assets")
            val asset = (0 until (assets?.length() ?: 0)).map { assets!!.getJSONObject(it) }.firstOrNull { it.optString("name") == ASSET }
            val want = asset?.optString("digest").orEmpty().removePrefix("sha256:").lowercase()
            if (asset == null || want.isEmpty()) Result.Failed("Update $latest has no APK with a published checksum")
            else {
                dir.mkdirs()
                dir.listFiles()?.forEach { it.delete() }
                val file = File(dir, "RichPresence-$latest.apk")
                val md = MessageDigest.getInstance("SHA-256")
                get(asset.getString("browser_download_url")).use { input ->
                    file.outputStream().use { out ->
                        val buf = ByteArray(1 shl 16)
                        while (true) {
                            val n = input.read(buf)
                            if (n < 0) break
                            md.update(buf, 0, n)
                            out.write(buf, 0, n)
                        }
                    }
                }
                val got = md.digest().joinToString("") { "%02x".format(it) }
                if (got != want) { file.delete(); Result.Failed("Update $latest didn't match its checksum and was deleted") }
                else Result.Ready(latest, file)
            }
        }
    } catch (e: Exception) {
        Result.Failed("Couldn't check for updates: ${e.message}")
    }

    private fun get(url: String) = (URL(url).openConnection() as HttpURLConnection).run {
        instanceFollowRedirects = true
        connectTimeout = 20_000
        readTimeout = 120_000
        setRequestProperty("User-Agent", "RichPresence updater")
        inputStream
    }

    /** Opens Android's installer for the downloaded APK (asking for the one-time permission first if needed). */
    fun install(a: Activity, file: File) {
        if (Build.VERSION.SDK_INT >= 26 && !a.packageManager.canRequestPackageInstalls()) {
            a.startActivity(Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES, Uri.parse("package:${a.packageName}")))
            return
        }
        val uri = FileProvider.getUriForFile(a, "${a.packageName}.files", file)
        a.startActivity(Intent(Intent.ACTION_VIEW).setDataAndType(uri, "application/vnd.android.package-archive")
            .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_ACTIVITY_NEW_TASK))
    }
}
